client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java

```diff
@@ -0,0 +1,306 @@
+                        } else {
+                            log.warn("Error in reading from socket.", e);
+                        }
+                        stop();
+                    } catch (EOFException e) {
+                        log.info("Closing TcpConnection.Reader because end of input readched.");
+                        stop();
+                    } catch (Exception e) {
+                        log.warn("Error processing data from from server " + name, e);
+                        stop();
+                    }
+                }
+            });
+        }
+
+        @VisibleForTesting
+        static WireCommand readCommand(InputStream in, IoBuffer buffer) throws IOException {
+            ByteBuf header = buffer.getBuffOfSize(in, 8);
+
+            int t = header.getInt(0);
+            WireCommandType type = WireCommands.getType(t);
+            if (type == null) {
+                throw new InvalidMessageException("Unknown wire command: " + t);
+            }
+
+            int length = header.getInt(4);
+            if (length < 0 || length > WireCommands.MAX_WIRECOMMAND_SIZE) {
+                throw new InvalidMessageException("Event of invalid length: " + length);
+            }
+
+            ByteBuf payload = buffer.getBuffOfSize(in, length);
+            
+            return type.readFrom(new EnhancedByteBufInputStream(payload), length);
+        }
+        
+        public void stop() {
+            stop.set(true);
+            thread.shutdown();
+            callback.connectionDropped();
+        }
+    }
+    
+    @RequiredArgsConstructor
+    private static final class TimeoutBatch implements Runnable {
+        private final AtomicLong token = new AtomicLong(-1);
+        private final CommandEncoder encoder;
+        @Override
+        public void run() {
+            token.set(encoder.batchTimeout(token.get()));
+        }    
+    }
+
+    public static CompletableFuture<TcpClientConnection> connect(PravegaNodeUri location, ClientConfig clientConfig, ReplyProcessor callback,
+                                              ScheduledExecutorService executor, Runnable onClose) {
+        return CompletableFuture.supplyAsync(() -> {
+            Socket socket = createClientSocket(location, clientConfig); 
+            try {
+                InputStream inputStream = socket.getInputStream();
+                AppendBatchSizeTrackerImpl batchSizeTracker = new AppendBatchSizeTrackerImpl();
+                ConnectionReader reader = new ConnectionReader(location.toString(), inputStream, callback, batchSizeTracker);
+                reader.start();
+                CommandEncoder encoder = new CommandEncoder(l -> batchSizeTracker, null, socket.getOutputStream());
+                return new TcpClientConnection(socket, encoder, reader, location, onClose, executor);
+            } catch (IOException e) {
+                try {
+                    onClose.run();
+                    socket.close();
```

Comment:
Should we ensure that the socket is closed before invoking onClose (as it may itself throw an exception)?

Reply:
I would close the socket on any exception that comes out of this code, not just IOException.

Reply:
Done

Reply:
It looks like that if `onClose` throws a RuntimeException we are still not calling `close`

Reply:
@eolivelli I think you need to refresh. close is now called first.

Reply:
I still see 
```
                    onClose.run();
                    socket.close();
```

Reply:
My mistake. I was looking in the wrong place.
fixed.